src/mame/drivers/gi6809.cpp

```diff
@@ -0,0 +1,1039 @@
+
+
+  * Game Jester Chance:
+
+  Incomplete due to a lack of hardware info and a pre-loaded nvram.
+  System runs partially.
+  Driver hasn't Coin In system hooked up.
+  Unknown NMI source on Master Cpu.
+
+
+****************************************************************************************
+
+  To Do:
+  ------
+
+  Discover the proper way to tie PIA's signals to Microprocessor's IRQs.
+
+
+****************************************************************************************/
+
+// Debug Notes:
+// good luck credits vars 6Bh - 1ABh - 501Bh(nvram) - bin
+// jester ch credits vars 34h - 179h - 5018h(nvram) - bin
+// castaway  credits vars 1Eh - 14Ah - 5025h(nvram) - bin
+
+#include "emu.h"
+
+#include "cpu/m6809/m6809.h"
+#include "machine/6821pia.h"
+#include "machine/nvram.h"
+#include "machine/ticket.h"
+#include "sound/discrete.h"
+#include "video/mc6845.h"
+
+#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+#include "tilemap.h"
+
+
+namespace {
+
+#define MASTER_CLOCK    XTAL(10'000'000)
+#define CPU_CLOCK_GL    (MASTER_CLOCK / 16)
+#define CPU_CLOCK       (MASTER_CLOCK / 2)
+#define PIXEL_CLOCK     (MASTER_CLOCK / 16)
+
+
+class gi6809_state : public driver_device
+{
+public:
+	gi6809_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_slavecpu(*this, "slavecpu"),
+		m_pia(*this, "pia%u", 0U),
+		m_screen(*this, "screen"),
+		m_gfxdecode(*this, "gfxdecode"),
+		m_discrete(*this, "discrete"),
+		m_videoram(*this, "videoram"),
+		m_colorram(*this, "colorram"),
+		m_input(*this, "IN0-%u", 0U),
+		m_coin_gl(*this, "CoinGL"),
+		m_coin_ca(*this, "CoinCA"),
+		m_hopper(*this, "hopper"),
+		m_lamps(*this, "lamp%u", 0U)
+	{ }
+
+	void gi6809_base(machine_config &config);
+	void castawayt(machine_config &config);
+	void glck6809(machine_config &config);
+	void jesterch(machine_config &config);
+	void init_cast();
+
+protected:
+	virtual void machine_start() override { m_lamps.resolve(); }
+	virtual void machine_reset() override;
+
+	virtual void video_start() override;
+
+private:
+	// Address Maps
+	void glckmain_map(address_map &map);
+	void glckslave_map(address_map &map);
+	void castmain_map(address_map &map);
+	void castslave_map(address_map &map);
+	void jestmain_map(address_map &map);
+	void jestslave_map(address_map &map);
+
+	// Devices
+	required_device<cpu_device> m_maincpu;
+	required_device<cpu_device> m_slavecpu;
+	required_device_array<pia6821_device, 2> m_pia;
+	required_device<screen_device> m_screen;
+	required_device<gfxdecode_device> m_gfxdecode;
+	required_device<discrete_device> m_discrete;
+	required_shared_ptr<uint8_t> m_videoram;
+	required_shared_ptr<uint8_t> m_colorram;
+	required_ioport_array<4> m_input;
+	optional_ioport m_coin_gl;
+	optional_ioport m_coin_ca;
+	optional_device<ticket_dispenser_device> m_hopper;
```

Comment:
Starting out with optional object finders is often a sign that you should split up your state class.